torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,534 @@
+  } else {
+    return vertex.get();
+  }
+}
+
+template <typename T>
+EdgeData<T> DynamicDAG<T>::removeVertex(Vertex<T>* v) {
+  for (auto* parent : v->in_edges()) {
+    EdgeData<T>::erase(parent->out_edges(), v);
+  }
+  for (auto* child : v->out_edges()) {
+    EdgeData<T>::erase(child->in_edges(), v);
+  }
+  auto edges = std::move(v->edges_);
+  vertices_[v->ord] = std::move(unique_vertex<T>(nullptr));
+  return edges;
+}
+
+/*
+ * [Edge addition]
+ * When adding an edge x -> y,
+ * - if ord(x) < ord(y), don't do anything.
+ * - if ord(y) < ord(x), some graph reordering must occur.
+ *
+ * Assume we are adding an edge x -> y and that ord(x) > ord(y).
+ * First, if there is a path y ----> x through some other vertices, then this
+ * edge addition would create a cycle. Figure this out via DFS and throw if necessary.
+ *
+ * Now, consider the set of all vertices v such that ord(x) > ord(v) > ord(y).
+ * Call this set the affected region (AR) -- these are the only vertices we
+ * need to consider for reordering to make the resulting graph valid.
+ *
+ * Find all children of y (through DFS) in AR (call this set deltaF and add y to it)
+ * Find all parents of x in AR (call this set deltaB and add x to it).
+ *
+ * Move y and all the children of y to after x and all the parents of x. The
+ * result topological ordering is valid.
+ *
+ * [Visual algorithm reference]
+ * Higher nodes come earlier in topological order.
+ * We are adding an edge between x -> y.
+ * The topological ordering is e, y, c, a, d, b, x, f.
+ * The affected region is {y, c, a, d, b, x}. e and f cannot be involved
+ * in the reorder.
+ *
+ *           (e)       <- ord = 0 ->                   (e)
+ *            |                                         |
+ *            v                                         v
+ *           (y)       <- ord = 1 ->        \          (c)
+ *            ^ \                       -----\          |
+ *        (c) |  v     <- ord = 2 ->    -----/     (d)  v
+ *          \ | (a)    <- ord = 3 ->        /       \->(x)
+ *           ||  |                                      /\
+ *       (d) ||  |     <- ord = 4 ->              (y)<-/  \
+ *        |  ||  v                                 \       |
+ *        \  v| (b)    <- ord = 5 ->                \->(a) |
+ *         ->(x)       <- ord = 6 ->             (b)<--/   v
+ *            \->(f)   <- ord = 7 ->                      (f)
+ *
+ * We find all children of y in the affected region. deltaF = {y, a, b}
+ * We find all parents of x via DFS. deltaB = {c, d, x}
+ *
+ * Now, we reorder all vertices in deltaB to come before deltaF.
+ * result = {c, d, x, y, a, b}, and we assign these {1, 2, 3, 4, 5, 6}.
+ *
+ * [Analysis]
+ * This is O(|AR|) which is the same thing as O(ord(consumer) - ord(producer))
+ * AR is the "affected region": { v s.t. ord(v) in [ord(producer), ord(consumer)] }
+ * consisting of the only verticies that can possibly be moved around due
+ * to this edge addition.
```

Comment:
It would be nice if we could say how the reordering happens, because it's a fairly non-trivial part of the algorithm (although very simple once you understand it).